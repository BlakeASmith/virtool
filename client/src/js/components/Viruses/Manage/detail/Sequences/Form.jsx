/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports SequenceForm
 */

'use strict';

var React = require('react');
var ReactDOM = require('react-dom');

var Row = require('react-bootstrap/lib/Row');
var Col = require('react-bootstrap/lib/Col');
var Input = require('react-bootstrap/lib/InputGroup');
var Overlay = require('react-bootstrap/lib/Overlay');
var Popover = require('react-bootstrap/lib/Popover');

var Icon = require('virtool/js/components/Base/Icon.jsx');
var PushButton = require('virtool/js/components/Base/PushButton.jsx');
var SequenceField = require('./SequenceField.jsx');


/**
 * A regular expression object used to test if a string contains spaces. Used to make sure the there are no spaces in
 * a supplied accession.
 *
 * @type {RegExp}
 * @object
 */
var SpaceRegEx = new RegExp(' ');


/**
 * A form-based component that is used for adding, editing, and reading sequence records. When reading, the form
 * cannot be modified. Editing can be toggled on a readOnly form and changes sent to the server.
 *
 * @class
 */
var SequenceForm = React.createClass({

    propTypes: {
        // Data from the sequence document that is used to populate the form. These props will be undefined when the form
        // is being used to add a new sequence.
        sequenceId: React.PropTypes.string,
        definition: React.PropTypes.string,
        host: React.PropTypes.string,
        sequence: React.PropTypes.string,

        // Callback from the parent prop to handle events in the form.
        update: React.PropTypes.func,
        onSubmit: React.PropTypes.func,
        autofill: React.PropTypes.func,

        mode: React.PropTypes.string,

        // Error message to display on the accession field.
        error: React.PropTypes.string
    },

    getDefaultProps: function () {
        return {
            sequenceId: '',
            definition: '',
            host: '',
            sequence: '',

            mode: "read"
        };
    },

    getInitialState: function () {
        return {
            pendingAutofill: false,
            pendingSave: false
        };
    },

    componentDidUpdate: function (prevProps) {
        if (!this.props.active && prevProps.active) {
            if (this.props.mode === "edit") this.refs.host.getInputDOMNode().focus();
            if (this.props.mode === "add") this.refs.accession.getInputDOMNode().focus();
        }
    },

    /**
     * Calls the onChange prop function with the input change object. Triggered by a change event in any of the form
     * input elements.
     *
     * @param event {object} - the input change event.
     * @func
     */
    handleChange: function (event) {
        var data = {};
        data[event.target.name] = event.target.value;
        this.props.update(data);
    },

    onSubmit: function (event) {
        event.preventDefault();
        console.log("submit");
    },

    render: function () {
        // An addon for the accession field that allows the user to pull data for the virus accession from NCBI.
        var autofillAddon;

        var overlay;

        if (this.props.mode === "add") {
            autofillAddon = (
                <PushButton onClick={this.props.autofill}>
                    <Icon name='wand' />
                </PushButton>
            );

            if (this.props.error) {
                // Set up an overlay to display if there is an error in state.
                var overlayProps = {
                    target: function () {return this.refs.accession.getInputDOMNode()}.bind(this),
                    placement: 'top'
                };

                overlay = (
                    <Overlay {...overlayProps} show={true} onHide={this.dismissError}>
                        <Popover id='sequence-error-popover'>
                            {this.props.error}
                        </Popover>
                    </Overlay>
                );
            }
        }

        var sharedProps = {
            spellCheck: false,
            readOnly: this.props.mode === "read",
            onChange: this.handleChange
        };

        return (
            <form onSubmit={this.onSubmit}>
                {overlay}

                <Row>
                    <Col md={6}>
                        <Input
                            ref='accession'
                            type='text'
                            name='sequenceId'
                            label='Accession'
                            value={this.props.sequenceId}
                            buttonAfter={autofillAddon}
                            {...sharedProps}
                            readOnly={this.props.mode !== "add"}
                        />
                    </Col>
                    <Col md={6}>
                        <Input
                            type='text'
                            name='host'
                            label='Host'
                            ref='host'
                            value={this.props.host}
                            placeholder={this.props.mode === "edit" ? "eg. Ageratum conyzoides": ""}
                            {...sharedProps}
                        />
                    </Col>
                </Row>
                <Row>
                    <Col md={12}>
                        <Input
                            type='text'
                            name='definition'
                            ref='definition'
                            label='Definition'
                            value={this.props.definition}
                            placeholder='eg. Ageratum enation virus, complete genome'
                            {...sharedProps}
                        />
                    </Col>
                </Row>
                <Row>
                    <Col md={12}>
                        <SequenceField
                            ref='sequence'
                            sequence={this.props.sequence}
                            {...sharedProps}
                        />
                    </Col>
                </Row>
            </form>
        );
    }
});

module.exports = SequenceForm;
