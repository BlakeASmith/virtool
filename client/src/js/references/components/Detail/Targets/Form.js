import React from "react";
import styled from "styled-components";
import { Input } from "../../../../base";

const StyledRequire = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;
export const TargetForm = ({ onChange, name, description, length, required }) => {
    return (
        <div>
            <Input label="Name" name="name" value={name} onChange={onChange} />
            <Input label="Description" name="description" value={description} onChange={onChange} />
            <Input type="number" label="Length" name="length" value={length} onChange={onChange} />
            <StyledRequire>
                <input type="checkbox" name="required" checked={required} onChange={onChange} />
                Required
            </StyledRequire>
        </div>
    );
};
