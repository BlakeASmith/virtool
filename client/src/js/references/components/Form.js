import React from "react";
import styled from "styled-components";
import { InputError } from "../../base";

const Name = styled.div`
    display: flex;
    flex-direction: column;
    width: ${props => (props.mode == "create" ? "49%" : "100%")};
`;

const Organism = styled.div`
    display: flex;
    flex-direction: column;
    width: 49%;
`;

const NameOrganism = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`;

export const ReferenceForm = ({ errorFile, errorSelect, name, onChange, errorName, description, organism, mode }) => {
    let extraComponent;
    let inputOrganism;

    if (errorFile != null || errorSelect != null) {
        extraComponent = (
            <div className="input-form-error">
                <span className="input-error-message" style={{ margin: "0 0 0 0" }}>
                    {errorFile || errorSelect}
                </span>
            </div>
        );
    }
    if (mode == "create") {
        inputOrganism = (
            <Organism>
                <InputError label="Organism" name="organism" value={organism} onChange={onChange} />
            </Organism>
        );
    }

    return (
        <div>
            <div>{extraComponent}</div>
            <NameOrganism>
                <Name mode={mode}>
                    <InputError label="Name" name="name" value={name} onChange={onChange} error={errorName} />
                </Name>
                {inputOrganism}
            </NameOrganism>

            <InputError
                label="Description"
                type="textarea"
                name="description"
                value={description}
                onChange={onChange}
            />
        </div>
    );
};
