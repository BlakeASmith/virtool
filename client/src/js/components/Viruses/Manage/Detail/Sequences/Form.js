/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports SequenceForm
 */

import React from "react";
import { Row, Col, Overlay, Popover, FormGroup, ControlLabel, FormControl, InputGroup } from "react-bootstrap";
import { Icon, Input, Button, SequenceField } from "virtool/js/components/Base";

/**
 * A form-based component that is used for adding, editing, and reading sequence records. When reading, the form
 * cannot be modified. Editing can be toggled on a readOnly form and changes sent to the server.
 *
 * @class
 */
export default class SequenceForm extends React.Component {

    constructor (props) {
        super(props);
        this.state = {
            pendingAutofill: false,
            pendingSave: false
        };
    }

    static propTypes = {
        // Data from the sequence document that is used to populate the form. These props will be undefined when the
        // form is being used to add a new sequence.
        sequenceId: React.PropTypes.string,
        definition: React.PropTypes.string,
        host: React.PropTypes.string,
        sequence: React.PropTypes.string,

        // Callback from the parent prop to handle events in the form.
        update: React.PropTypes.func,
        onSubmit: React.PropTypes.func,
        autofill: React.PropTypes.func,

        active: React.PropTypes.bool,
        mode: React.PropTypes.string,

        // Error message to display on the accession field.
        error: React.PropTypes.string
    };

    static defaultProps = () => {
        return {
            sequenceId: "",
            definition: "",
            host: "",
            sequence: "",
            mode: "read"
        };
    };

    componentDidUpdate (prevProps) {
        if (!this.props.active && prevProps.active) {
            if (this.props.mode === "edit") {
                this.hostNode.focus();
            }
            if (this.props.mode === "add") {
                this.accessionNode.focus();
            }
        }
    }

    /**
     * Calls the onChange prop function with the input change object. Triggered by a change event in any of the form
     * input elements.
     *
     * @param event {object} - the input change event.
     * @func
     */
    handleChange = (event) => {
        let data = {};
        data[event.target.name] = event.target.value;
        this.props.update(data);
    };

    onSubmit = (event) => {
        event.preventDefault();
    };

    render () {
        let overlay;

        const sharedProps = {
            spellCheck: false,
            readOnly: this.props.mode === "read",
            onChange: this.handleChange
        };

        let accession = (
            <FormControl
                {...sharedProps}
                inputRef={(node) => this.accessionNode = node}
                type="text"
                name="sequenceId"
                value={this.props.sequenceId}
                readOnly={this.props.mode !== "add"}
            />
        );

        if (this.props.mode === "add") {
            accession = (
                <InputGroup>
                    {accession}
                    <InputGroup.Button>
                        <Button onClick={this.props.autofill}>
                            <Icon name="wand" />
                        </Button>
                    </InputGroup.Button>
                </InputGroup>
            );

            if (this.props.error) {
                // Set up an overlay to display if there is an error in state.
                const overlayProps = {
                    target: this.accessionNode,
                    placement: "top"
                };

                overlay = (
                    <Overlay {...overlayProps} show={true} onHide={this.dismissError}>
                        <Popover id="sequence-error-popover">
                            {this.props.error}
                        </Popover>
                    </Overlay>
                );
            }
        }

        return (
            <form onSubmit={this.onSubmit}>
                {overlay}

                <Row>
                    <Col md={6}>
                        <FormGroup>
                            <ControlLabel>Accession</ControlLabel>
                            {accession}
                        </FormGroup>
                    </Col>
                    <Col md={6}>
                        <Input
                            type="text"
                            name="host"
                            label="Host"
                            value={this.props.host}
                            placeholder={this.props.mode === "edit" ? "eg. Ageratum conyzoides": ""}
                            {...sharedProps}
                        />
                    </Col>
                </Row>
                <Row>
                    <Col md={12}>
                        <Input
                            type="text"
                            name="definition"
                            label="Definition"
                            value={this.props.definition}
                            placeholder="eg. Ageratum enation virus, complete genome"
                            {...sharedProps}
                        />
                    </Col>
                </Row>
                <Row>
                    <Col md={12}>
                        <SequenceField
                            sequence={this.props.sequence}
                            {...sharedProps}
                        />
                    </Col>
                </Row>
            </form>
        );
    }
}
