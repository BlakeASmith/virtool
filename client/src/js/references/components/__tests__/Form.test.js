import { ReferenceForm } from "../Form";

describe("<ReferenceForm />", () => {
    let props;
    beforeEach(() => {
        props = {
            errorFile: "foo",
            errorSelect: "bar",
            name: "fee",
            onChange: jest.fn(),
            errorName: "baz",
            description: "Foo",
            organism: "Bar",
            mode: "clone"
        };
    });

    it("should render when [errorFile!=null]", () => {
        const wrapper = shallow(<ReferenceForm {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should render when [errorFile=null] and [errorSelect=null]", () => {
        props.errorFile = null;
        props.errorSelect = null;
        const wrapper = shallow(<ReferenceForm {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
    it("should render when [mode=create]", () => {
        props.mode = "create";
        const wrapper = shallow(<ReferenceForm {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});
