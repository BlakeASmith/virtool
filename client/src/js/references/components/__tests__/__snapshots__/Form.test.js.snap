// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReferenceForm /> should render when [errorFile!=null] 1`] = `
<div>
  <div>
    <div
      className="input-form-error"
    >
      <span
        className="input-error-message"
        style={
          Object {
            "margin": "0 0 0 0",
          }
        }
      >
        foo
      </span>
    </div>
  </div>
  <Form__NameOrganism>
    <Form__Name
      mode="clone"
    >
      <InputError
        error="baz"
        label="Name"
        name="name"
        onChange={[MockFunction]}
        value="fee"
      />
    </Form__Name>
  </Form__NameOrganism>
  <InputError
    label="Description"
    name="description"
    onChange={[MockFunction]}
    type="textarea"
    value="Foo"
  />
</div>
`;

exports[`<ReferenceForm /> should render when [errorFile=null] and [errorSelect=null] 1`] = `
<div>
  <div />
  <Form__NameOrganism>
    <Form__Name
      mode="clone"
    >
      <InputError
        error="baz"
        label="Name"
        name="name"
        onChange={[MockFunction]}
        value="fee"
      />
    </Form__Name>
  </Form__NameOrganism>
  <InputError
    label="Description"
    name="description"
    onChange={[MockFunction]}
    type="textarea"
    value="Foo"
  />
</div>
`;

exports[`<ReferenceForm /> should render when [mode=create] 1`] = `
<div>
  <div>
    <div
      className="input-form-error"
    >
      <span
        className="input-error-message"
        style={
          Object {
            "margin": "0 0 0 0",
          }
        }
      >
        foo
      </span>
    </div>
  </div>
  <Form__NameOrganism>
    <Form__Name
      mode="create"
    >
      <InputError
        error="baz"
        label="Name"
        name="name"
        onChange={[MockFunction]}
        value="fee"
      />
    </Form__Name>
    <Form__Organism>
      <InputError
        label="Organism"
        name="organism"
        onChange={[MockFunction]}
        value="Bar"
      />
    </Form__Organism>
  </Form__NameOrganism>
  <InputError
    label="Description"
    name="description"
    onChange={[MockFunction]}
    type="textarea"
    value="Foo"
  />
</div>
`;
